## 1. Import and Explore

In [177]:
import pandas as pd

df_marketing_spend = pd.read_csv('marketing_spend.csv')
df_orders = pd.read_csv('orders.csv')
df_products = pd.read_csv('products.csv')


In [178]:
df_marketing_spend

,month,platform,spend,impressions,clicks,conversions,revenue_attributed,cpc,cpa,roas
0,2024-01,Google Ads,2676.14,303594,13271,500,43063.16,0.20,5.35,16.09
1,2024-01,Facebook Ads,1971.43,338022,7163,333,15950.86,0.28,5.92,8.09
2,2024-01,Instagram Ads,2418.95,561361,18729,748,54454.46,0.13,3.23,22.51
3,2024-01,TikTok Ads,1141.70,554273,6967,346,22369.23,0.16,3.30,19.59
4,2024-01,Email Marketing,868.82,29509,1215,58,4223.97,0.72,14.98,4.86
...,...,...,...,...,...,...,...,...,...,...
139,2025-12,Facebook Ads,7480.08,1454093,34543,2668,201086.58,0.22,2.80,26.88
140,2025-12,Instagram Ads,5671.66,910841,21625,869,91175.48,0.26,6.53,16.08
141,2025-12,TikTok Ads,3408.50,755034,18046,1171,111842.60,0.19,2.91,32.81
142,2025-12,Email Marketing,1796.66,15675,477,11,1198.99,3.77,163.33,0.67


In [179]:
df_orders

,order_id,customer_id,order_date,channel,payment_method,region,items_ordered,primary_category,gross_revenue,discount_pct,discount_amount,shipping_cost,product_cost,platform_fee,transaction_fee,returned,refund_amount,net_revenue,total_costs,profit
0,ORD-000001,C-0617,2025-06-24,Mobile App,Gift Card,Northeast,2,Electronics,261.39,0,0.00,22.28,92.30,0.00,7.88,No,0.00,261.39,122.46,138.93
1,ORD-000002,C-0720,2025-08-23,Website,Credit Card,Midwest,1,Beauty,24.02,0,0.00,3.98,10.75,0.00,1.00,No,0.00,24.02,15.73,8.29
2,ORD-000003,C-0189,2025-07-20,Website,Credit Card,Southeast,2,Toys,52.60,0,0.00,14.60,31.41,0.00,1.83,No,0.00,52.60,47.84,4.76
3,ORD-000004,C-0058,2025-07-17,Mobile App,Debit Card,Southeast,2,Sports,30.63,5,1.53,30.98,16.27,0.00,1.14,No,0.00,29.10,48.39,-19.29
4,ORD-000005,C-0580,2025-10-11,Marketplace,Credit Card,Northeast,1,Sports,11.99,25,3.00,10.30,6.88,1.35,0.56,No,0.00,8.99,19.09,-10.10
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1995,ORD-001996,C-0250,2025-09-06,Mobile App,Debit Card,Midwest,2,Electronics,178.22,0,0.00,51.04,80.51,0.00,5.47,No,0.00,178.22,137.02,41.20
1996,ORD-001997,C-0652,2025-06-27,Mobile App,Buy Now Pay Later,Midwest,3,Beauty,201.02,0,0.00,42.54,75.39,0.00,6.13,No,0.00,201.02,124.06,76.96
1997,ORD-001998,C-0473,2025-01-20,Marketplace,Credit Card,Southeast,1,Beauty,35.41,0,0.00,12.14,18.89,5.31,1.33,Yes,35.41,0.00,37.67,-37.67
1998,ORD-001999,C-0784,2025-08-14,Website,Credit Card,Northeast,2,Toys,81.67,10,8.17,27.08,36.91,0.00,2.43,No,0.00,73.50,66.42,7.08


In [180]:
df_products

,product_id,product_name,category,sub_category,unit_cost,selling_price,shipping_cost_per_unit,weight_lbs,supplier
0,SKU-0001,Headphones - Variant 1,Electronics,Headphones,29.44,56.08,13.42,12.4,Supplier A
1,SKU-0002,Headphones - Variant 2,Electronics,Headphones,26.92,91.57,15.50,15.0,Supplier B
2,SKU-0003,Headphones - Variant 3,Electronics,Headphones,15.91,43.34,9.58,7.6,Supplier C
3,SKU-0004,Headphones - Variant 4,Electronics,Headphones,7.13,24.31,5.74,2.8,Supplier B
4,SKU-0005,Headphones - Variant 5,Electronics,Headphones,43.05,135.82,5.82,2.9,Supplier C
...,...,...,...,...,...,...,...,...,...
202,SKU-0203,Sauces - Variant 5,Food & Beverage,Sauces,19.99,32.04,14.22,13.4,Supplier C
203,SKU-0204,Organic - Variant 1,Food & Beverage,Organic,13.64,27.16,11.34,9.8,Supplier B
204,SKU-0205,Organic - Variant 2,Food & Beverage,Organic,8.47,19.41,12.70,11.5,Supplier D
205,SKU-0206,Organic - Variant 3,Food & Beverage,Organic,39.68,136.04,6.30,3.5,Supplier D


In [181]:
df_orders.columns

Index(['order_id', 'customer_id', 'order_date', 'channel', 'payment_method',
       'region', 'items_ordered', 'primary_category', 'gross_revenue',
       'discount_pct', 'discount_amount', 'shipping_cost', 'product_cost',
       'platform_fee', 'transaction_fee', 'returned', 'refund_amount',
       'net_revenue', 'total_costs', 'profit'],
      dtype='object')

In [182]:
#Order
import numpy as np
total_cost = df_orders['product_cost'] + df_orders['shipping_cost'] + df_orders['platform_fee'] + df_orders['transaction_fee']

mask = ~np.isclose(total_cost,df_orders['total_costs'])
df_mismatch = df_orders[mask]
df_mismatch.head(5)


,order_id,customer_id,order_date,channel,payment_method,region,items_ordered,primary_category,gross_revenue,discount_pct,discount_amount,shipping_cost,product_cost,platform_fee,transaction_fee,returned,refund_amount,net_revenue,total_costs,profit


## 2. Category Profitability

In [183]:
df_orders['primary_category'].unique()

array(['Electronics', 'Beauty', 'Toys', 'Sports', 'Clothing',
       'Food & Beverage', 'Books', 'Home & Kitchen'], dtype=object)

In [184]:
df_orders['profit_margin'] = np.where(df_orders['gross_revenue'] != 0,df_orders['profit'] / df_orders['gross_revenue'] * 100, np.nan)

df_orders_category = (df_orders
                      .groupby('primary_category', dropna = False, as_index = False)
                      .agg(
                            profit = ('profit', 'sum'),
                            profit_margin = ('profit_margin','sum'),
                            gross_revenue = ('gross_revenue','sum'),
                            total_costs = ('total_costs','sum'),
                            category_size = ('primary_category', 'count'),
                    )
)

In [185]:
df_orders_category.sort_values('profit', ascending=False)

,primary_category,profit,profit_margin,gross_revenue,total_costs,category_size
3,Electronics,13973.46,4494.868332,52876.49,30912.38,267
7,Toys,8786.25,3058.716363,39081.00,24809.43,257
6,Sports,7596.63,2567.681813,37453.25,24733.20,292
4,Food & Beverage,7591.77,3053.660227,36878.52,23070.92,247
5,Home & Kitchen,6688.57,3301.388256,30879.43,19671.19,200
2,Clothing,6272.48,1922.487367,37515.60,25110.65,293
0,Beauty,3174.88,-466.426870,20871.36,15079.19,205
1,Books,2249.79,-3050.675311,22413.48,16597.58,239


## 3. Channel Analysis

In [186]:
df_orders['channel'].unique()

array(['Mobile App', 'Website', 'Marketplace', 'Social Commerce'],
      dtype=object)

In [187]:

df_orders_channel = (df_orders
                     .groupby('channel',dropna = False, as_index=False)
                     .agg(
                        average_profit=('profit','mean'),
                        profit_margin = ('profit_margin','sum'),
                        gross_revenue = ('gross_revenue','sum'),
                        total_costs = ('total_costs','sum'),
                        platform_fee = ('platform_fee', 'sum'),
                        total_orders = ('channel','count')
                          )
                     )
df_orders_channel['average_order_value'] = df_orders_channel['gross_revenue'] / df_orders_channel['total_orders'] 

In [188]:
df_orders_channel.sort_values('average_profit',ascending=False)

,channel,average_profit,profit_margin,gross_revenue,total_costs,platform_fee,total_orders,average_order_value
1,Mobile App,36.321392,6953.379826,82725.94,50500.09,0.00,589,140.451511
3,Website,31.595547,8509.982208,111165.14,67872.09,0.00,795,139.830365
2,Social Commerce,17.109340,460.431983,26446.61,18558.83,1944.93,197,134.246751
0,Marketplace,15.397446,-1042.093842,57631.44,43053.53,7949.69,419,137.545203


In [189]:
df_orders_channel.sort_values('platform_fee', ascending = False)

,channel,average_profit,profit_margin,gross_revenue,total_costs,platform_fee,total_orders,average_order_value
0,Marketplace,15.397446,-1042.093842,57631.44,43053.53,7949.69,419,137.545203
2,Social Commerce,17.109340,460.431983,26446.61,18558.83,1944.93,197,134.246751
1,Mobile App,36.321392,6953.379826,82725.94,50500.09,0.00,589,140.451511
3,Website,31.595547,8509.982208,111165.14,67872.09,0.00,795,139.830365


# 4. Marketing ROI

CPC = Cost Per Click (How much you pay for each click)<br>
$$
CPC = \frac{Spend}{Clicks}
$$

CPA = Cost Per Acquistion (How much costs to generate one order/customer) <br>
$$
CPA = \frac{Spend}{Conversions}
$$

CTR = Click Through Rate (How many people clicked after seeing the ad)
$$
CTR = \frac{Clicks}{Impression}
$$

CVR = Conversion Rate (How many people clicks become customer)
$$
CVR = \frac{Conversions}{Clicks}
$$

RPC = Revenue Per Clicks (How many revenue each click generates)
$$
RPC = \frac{Revenue}{Clicks}
$$

ROAS = Return On Ad Spend (Every 1 dollar generates 4 dollars revenue)
$$
ROAS = \frac{Revenue}{Spend}
$$


In [190]:
df_marketing_spend

,month,platform,spend,impressions,clicks,conversions,revenue_attributed,cpc,cpa,roas
0,2024-01,Google Ads,2676.14,303594,13271,500,43063.16,0.20,5.35,16.09
1,2024-01,Facebook Ads,1971.43,338022,7163,333,15950.86,0.28,5.92,8.09
2,2024-01,Instagram Ads,2418.95,561361,18729,748,54454.46,0.13,3.23,22.51
3,2024-01,TikTok Ads,1141.70,554273,6967,346,22369.23,0.16,3.30,19.59
4,2024-01,Email Marketing,868.82,29509,1215,58,4223.97,0.72,14.98,4.86
...,...,...,...,...,...,...,...,...,...,...
139,2025-12,Facebook Ads,7480.08,1454093,34543,2668,201086.58,0.22,2.80,26.88
140,2025-12,Instagram Ads,5671.66,910841,21625,869,91175.48,0.26,6.53,16.08
141,2025-12,TikTok Ads,3408.50,755034,18046,1171,111842.60,0.19,2.91,32.81
142,2025-12,Email Marketing,1796.66,15675,477,11,1198.99,3.77,163.33,0.67


In [191]:
df_marketing_spend_platform = (df_marketing_spend
                               .groupby('platform', dropna=False, as_index=False)
                              .agg(
                                spend = ('spend','sum'),
                                impressions = ('impressions','sum'),
                                clicks = ('clicks','sum'),
                                conversions = ('conversions','sum'),
                                revenue_attributed = ('revenue_attributed','sum'),
                              )
                              )

df_marketing_spend_platform['ROAS'] =np.where(df_marketing_spend_platform['spend'] > 0 ,
                                             (df_marketing_spend_platform['revenue_attributed'] / df_marketing_spend_platform['spend']).round(2),
                                             np.nan)
df_marketing_spend_platform['CPC'] =np.where(df_marketing_spend_platform['clicks'] > 0 ,
                                             (df_marketing_spend_platform['spend'] / df_marketing_spend_platform['clicks']).round(2),
                                             np.nan)
df_marketing_spend_platform['CPA'] =np.where(df_marketing_spend_platform['conversions'] > 0 ,
                                             (df_marketing_spend_platform['spend'] / df_marketing_spend_platform['conversions']).round(2),
                                             np.nan)
df_marketing_spend_platform['CTR'] =np.where(df_marketing_spend_platform['impressions'] > 0 ,
                                             (df_marketing_spend_platform['spend'] / df_marketing_spend_platform['impressions']).round(3),
                                             np.nan)
df_marketing_spend_platform['CVR'] =np.where(df_marketing_spend_platform['clicks'] > 0 ,
                                             (df_marketing_spend_platform['conversions'] / df_marketing_spend_platform['clicks']).round(4),
                                             np.nan)
df_marketing_spend_platform['RPC'] =np.where(df_marketing_spend_platform['spend'] > 0 ,
                                             (df_marketing_spend_platform['revenue_attributed'] / df_marketing_spend_platform['spend']).round(2),
                                             np.nan)

df_marketing_spend_platform.sort_values('spend', ascending = False)

,platform,spend,impressions,clicks,conversions,revenue_attributed,ROAS,CPC,CPA,CTR,CVR,RPC
2,Google Ads,152546.48,17529367,611740,32309,2194120.83,14.38,0.25,4.72,0.009,0.0528,14.38
1,Facebook Ads,106451.93,16305609,331746,16968,1218572.43,11.45,0.32,6.27,0.007,0.0511,11.45
3,Influencer,97663.12,9990653,587501,28824,2216974.26,22.70,0.17,3.39,0.010,0.0491,22.70
4,Instagram Ads,65154.02,11373438,306843,13863,1024639.06,15.73,0.21,4.70,0.006,0.0452,15.73
5,TikTok Ads,57229.22,20750539,383269,18684,1374627.17,24.02,0.15,3.06,0.003,0.0487,24.02
0,Email Marketing,24461.37,636743,25971,1370,117681.45,4.81,0.94,17.86,0.038,0.0528,4.81


In [192]:
df_marketing_spend_platform['ctr'] = np.where(df_marketing_spend_platform['impressions'] > 0, 
                                             df_marketing_spend_platform['clicks']/df_marketing_spend_platform['impressions'],
                                              np.nan)
df_marketing_spend_platform['cvr'] = np.where(df_marketing_spend_platform['clicks'] > 0, 
                                             df_marketing_spend_platform['conversions']/df_marketing_spend_platform['clicks'],
                                              np.nan)
df_marketing_spend_platform['rpc'] = np.where(df_marketing_spend_platform['clicks'] > 0, 
                                             df_marketing_spend_platform['revenue_attributed']/df_marketing_spend_platform['clicks'],
                                              np.nan)

In [193]:
df_marketing_spend_platform.sort_values('spend', ascending = False)

,platform,spend,impressions,clicks,conversions,revenue_attributed,ROAS,CPC,CPA,CTR,CVR,RPC,ctr,cvr,rpc
2,Google Ads,152546.48,17529367,611740,32309,2194120.83,14.38,0.25,4.72,0.009,0.0528,14.38,0.034898,0.052815,3.586689
1,Facebook Ads,106451.93,16305609,331746,16968,1218572.43,11.45,0.32,6.27,0.007,0.0511,11.45,0.020346,0.051148,3.673209
3,Influencer,97663.12,9990653,587501,28824,2216974.26,22.70,0.17,3.39,0.010,0.0491,22.70,0.058805,0.049062,3.773567
4,Instagram Ads,65154.02,11373438,306843,13863,1024639.06,15.73,0.21,4.70,0.006,0.0452,15.73,0.026979,0.045179,3.339294
5,TikTok Ads,57229.22,20750539,383269,18684,1374627.17,24.02,0.15,3.06,0.003,0.0487,24.02,0.018470,0.048749,3.586586
0,Email Marketing,24461.37,636743,25971,1370,117681.45,4.81,0.94,17.86,0.038,0.0528,4.81,0.040787,0.052751,4.531264


In [194]:
df_marketing_spend_platform[['platform','spend']].sort_values('spend', ascending = False)

,platform,spend
2,Google Ads,152546.48
1,Facebook Ads,106451.93
3,Influencer,97663.12
4,Instagram Ads,65154.02
5,TikTok Ads,57229.22
0,Email Marketing,24461.37


In [195]:
import numpy as np
import pandas as pd

df_marketing_spend['month'] = pd.to_datetime(df_marketing_spend['month'])

df_marketing_monthly = (
    df_marketing_spend
    .groupby(['month', 'platform'], as_index=False)
    .agg(
        spend=('spend', 'sum'),
        impressions=('impressions', 'sum'),
        clicks=('clicks', 'sum'),
        conversions=('conversions', 'sum'),
        revenue_attributed=('revenue_attributed', 'sum')
    )
)

df_marketing_monthly['cpc'] = np.where(
    df_marketing_monthly['clicks'] > 0,
    df_marketing_monthly['spend'] / df_marketing_monthly['clicks'],
    np.nan
)

df_marketing_monthly['cpa'] = np.where(
    df_marketing_monthly['conversions'] > 0,
    df_marketing_monthly['spend'] / df_marketing_monthly['conversions'],
    np.nan
)

df_marketing_monthly['roas'] = np.where(
    df_marketing_monthly['spend'] > 0,
    df_marketing_monthly['revenue_attributed'] / df_marketing_monthly['spend'],
    np.nan
)

df_marketing_monthly['ctr'] = np.where(
    df_marketing_monthly['impressions'] > 0,
    df_marketing_monthly['clicks'] / df_marketing_monthly['impressions'],
    np.nan
)

df_marketing_monthly['cvr'] = np.where(
    df_marketing_monthly['clicks'] > 0,
    df_marketing_monthly['conversions'] / df_marketing_monthly['clicks'],
    np.nan
)

df_marketing_monthly['rpc'] = np.where(
    df_marketing_monthly['clicks'] > 0,
    df_marketing_monthly['revenue_attributed'] / df_marketing_monthly['clicks'],
    np.nan
)

df_marketing_monthly

,month,platform,spend,impressions,clicks,conversions,revenue_attributed,cpc,cpa,roas,ctr,cvr,rpc
0,2024-01-01,Email Marketing,868.82,29509,1215,58,4223.97,0.715078,14.979655,4.861732,0.041174,0.047737,3.476519
1,2024-01-01,Facebook Ads,1971.43,338022,7163,333,15950.86,0.275224,5.920210,8.091010,0.021191,0.046489,2.226841
2,2024-01-01,Google Ads,2676.14,303594,13271,500,43063.16,0.201653,5.352280,16.091520,0.043713,0.037676,3.244907
3,2024-01-01,Influencer,2229.89,324906,18550,672,32663.79,0.120210,3.318289,14.648162,0.057093,0.036226,1.760851
4,2024-01-01,Instagram Ads,2418.95,561361,18729,748,54454.46,0.129155,3.233890,22.511610,0.033364,0.039938,2.907494
...,...,...,...,...,...,...,...,...,...,...,...,...,...
139,2025-12-01,Facebook Ads,7480.08,1454093,34543,2668,201086.58,0.216544,2.803628,26.882945,0.023756,0.077237,5.821341
140,2025-12-01,Google Ads,10726.23,1557007,75253,2406,141924.78,0.142536,4.458117,13.231562,0.048332,0.031972,1.885968
141,2025-12-01,Influencer,6329.25,337538,17751,1374,103555.05,0.356557,4.606441,16.361346,0.052590,0.077404,5.833759
142,2025-12-01,Instagram Ads,5671.66,910841,21625,869,91175.48,0.262273,6.526651,16.075625,0.023742,0.040185,4.216207


In [196]:
df_roas_plot = df_marketing_monthly.pivot(
    index='month',
    columns='platform',
    values='roas'
)

df_roas_plot

platform,Email Marketing,Facebook Ads,Google Ads,Influencer,Instagram Ads,TikTok Ads
month,,,,,,
2024-01-01,4.861732,8.091010,16.091520,14.648162,22.511610,19.592914
2024-02-01,7.982224,22.990973,18.034549,47.918764,23.071559,39.597467
2024-03-01,16.646738,12.055900,27.929699,14.106765,26.029392,43.871961
2024-04-01,4.891397,8.452004,12.356053,5.831173,13.302710,12.144801
2024-05-01,7.256259,24.393376,10.840071,18.976481,45.290750,10.714589
2024-06-01,1.797208,16.689592,11.588582,33.460199,17.557002,24.957801
2024-07-01,7.632983,11.791699,9.085209,13.132912,7.922928,28.698821
2024-08-01,4.968345,3.094685,11.404345,25.349324,9.794853,6.014405
2024-09-01,3.892558,16.859470,24.171830,17.590604,37.330423,17.194991


# Question to Answer

1. What is the average profit margin by product category? Which categories are the most and least profitable, and what is driving the difference (product cost, shipping, returns, or discounts)?

In [199]:
df_order_category = (df_orders.groupby('primary_category', dropna = False, as_index = False)
                     .agg(
                         product_cost = ('product_cost','sum'),
                         shipping_cost = ('shipping_cost','sum'),
                         refund_amount = ('refund_amount','sum'),
                         discount_amount = ('discount_amount','sum'),
                         profit = ('profit','sum'),
                         gross_revenue = ('gross_revenue','sum')
                     ))
df_order_category['average_profit_margin'] = df_order_category['profit'] / df_order_category['gross_revenue'] * 100
df_order_category.sort_values('average_profit_margin', ascending = False)

,primary_category,product_cost,shipping_cost,refund_amount,discount_amount,profit,gross_revenue,average_profit_margin
3,Electronics,20091.06,7139.80,4078.27,3912.38,13973.46,52876.49,26.426603
7,Toys,15610.64,6740.36,2469.81,3015.51,8786.25,39081.00,22.482152
5,Home & Kitchen,12348.67,4988.84,2340.10,2179.57,6688.57,30879.43,21.660277
4,Food & Beverage,14299.07,6441.76,3504.56,2711.27,7591.77,36878.52,20.585886
6,Sports,15691.62,6914.72,2008.44,3114.98,7596.63,37453.25,20.282966
2,Clothing,15345.43,7444.60,3209.21,2923.26,6272.48,37515.60,16.719658
0,Beauty,8590.59,5172.98,865.70,1751.59,3174.88,20871.36,15.211658
1,Books,8956.87,6262.80,2106.36,1459.75,2249.79,22413.48,10.037665


Answer: The average profit margin varies across product categories, reflecting differences in cost structures such as product cost, shipping cost, refunds, and discounts. Based on the analysis, Electronics has the highest average profit margin at approximately 26.43%, making it the most profitable category. This strong performance is largely driven by its relatively high gross revenue ($52,876.49) compared to its combined costs, allowing it to maintain a larger profit of $13,973.46 despite moderate expenses in shipping, refunds, and discounts. Toys (22.48%) and Home & Kitchen (21.66%) also show strong profitability, indicating that their revenue levels sufficiently offset operational costs. In contrast, Books has the lowest average profit margin at about 10.04%, making it the least profitable category. This lower profitability appears to be driven primarily by high shipping costs ($6,262.80) relative to its gross revenue ($22,413.48), along with additional deductions from refunds and discounts that further reduce profit. Similarly, categories like Beauty and Clothing show lower margins due to a combination of relatively high product costs and operational expenses compared to their revenue. Overall, the variation in profitability across categories is mainly influenced by the balance between revenue and operational costs, particularly product costs and shipping expenses, while refunds and discounts further compress margins in several categories.

2. How does profitability differ across sales channels (Website, Mobile App, Marketplace, Social Commerce)? Which channel has the best and worst profit per order after accounting for platform fees?

In [200]:
df_orders_channel.sort_values('average_profit',ascending=False)

,channel,average_profit,profit_margin,gross_revenue,total_costs,platform_fee,total_orders,average_order_value
1,Mobile App,36.321392,6953.379826,82725.94,50500.09,0.00,589,140.451511
3,Website,31.595547,8509.982208,111165.14,67872.09,0.00,795,139.830365
2,Social Commerce,17.109340,460.431983,26446.61,18558.83,1944.93,197,134.246751
0,Marketplace,15.397446,-1042.093842,57631.44,43053.53,7949.69,419,137.545203


Answer: Profitability varies significantly across the different sales channels when considering profit per order and the impact of platform fees. The Mobile App channel shows the highest profitability, with an average profit per order of approximately $36.32, followed by the Website channel at about $31.60 per order. Both channels benefit from zero platform fees, allowing them to retain more revenue from each transaction. Additionally, they generate the highest total revenue and order volumes, which further supports their overall profitability.

In contrast, Social Commerce generates a lower average profit per order of about $17.11, partly due to the presence of platform fees totaling $1,944.93, which reduce the net profit from each order. The Marketplace channel performs the worst, with the lowest profit per order at approximately $15.40 and a negative profit margin overall, largely driven by high platform fees totaling $7,949.69. These fees significantly increase total costs relative to revenue, compressing margins and reducing profitability.

Overall, Mobile App and Website are the most profitable channels, primarily because they avoid platform fees and maintain higher profit per order, while Marketplace is the least profitable channel due to substantial platform fees that erode margins despite relatively strong revenue and order volume.

3. What is the return rate by category and channel? Estimate how much total revenue was lost to returns over the analysis period.

In [201]:
df_orders['returned_binary'] = df_orders['returned'].map({'Yes':1,'No':0})
df_orders

,order_id,customer_id,order_date,channel,payment_method,region,items_ordered,primary_category,gross_revenue,discount_pct,...,product_cost,platform_fee,transaction_fee,returned,refund_amount,net_revenue,total_costs,profit,profit_margin,returned_binary
0,ORD-000001,C-0617,2025-06-24,Mobile App,Gift Card,Northeast,2,Electronics,261.39,0,...,92.30,0.00,7.88,No,0.00,261.39,122.46,138.93,53.150465,0
1,ORD-000002,C-0720,2025-08-23,Website,Credit Card,Midwest,1,Beauty,24.02,0,...,10.75,0.00,1.00,No,0.00,24.02,15.73,8.29,34.512906,0
2,ORD-000003,C-0189,2025-07-20,Website,Credit Card,Southeast,2,Toys,52.60,0,...,31.41,0.00,1.83,No,0.00,52.60,47.84,4.76,9.049430,0
3,ORD-000004,C-0058,2025-07-17,Mobile App,Debit Card,Southeast,2,Sports,30.63,5,...,16.27,0.00,1.14,No,0.00,29.10,48.39,-19.29,-62.977473,0
4,ORD-000005,C-0580,2025-10-11,Marketplace,Credit Card,Northeast,1,Sports,11.99,25,...,6.88,1.35,0.56,No,0.00,8.99,19.09,-10.10,-84.236864,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1995,ORD-001996,C-0250,2025-09-06,Mobile App,Debit Card,Midwest,2,Electronics,178.22,0,...,80.51,0.00,5.47,No,0.00,178.22,137.02,41.20,23.117495,0
1996,ORD-001997,C-0652,2025-06-27,Mobile App,Buy Now Pay Later,Midwest,3,Beauty,201.02,0,...,75.39,0.00,6.13,No,0.00,201.02,124.06,76.96,38.284748,0
1997,ORD-001998,C-0473,2025-01-20,Marketplace,Credit Card,Southeast,1,Beauty,35.41,0,...,18.89,5.31,1.33,Yes,35.41,0.00,37.67,-37.67,-106.382378,1
1998,ORD-001999,C-0784,2025-08-14,Website,Credit Card,Northeast,2,Toys,81.67,10,...,36.91,0.00,2.43,No,0.00,73.50,66.42,7.08,8.669034,0


In [216]:

df_orders_channel = (df_orders
                     .groupby('channel',dropna = False, as_index=False)
                     .agg(
                        average_profit=('profit','mean'),
                        profit_margin = ('profit_margin','sum'),
                        gross_revenue = ('gross_revenue','sum'),
                        total_costs = ('total_costs','sum'),
                        platform_fee = ('platform_fee', 'sum'),
                        total_orders = ('channel','count'),
                        returned_binary = ('returned_binary', 'sum'),
                        refund_amount = ('refund_amount', 'sum')
                          )
                     )
df_orders_channel['average_order_value'] = df_orders_channel['gross_revenue'] / df_orders_channel['total_orders'] 

In [217]:
df_orders_channel

,channel,average_profit,profit_margin,gross_revenue,total_costs,platform_fee,total_orders,returned_binary,refund_amount,average_order_value
0,Marketplace,15.397446,-1042.093842,57631.44,43053.53,7949.69,419,27,3492.95,137.545203
1,Mobile App,36.321392,6953.379826,82725.94,50500.09,0.00,589,43,5324.17,140.451511
2,Social Commerce,17.109340,460.431983,26446.61,18558.83,1944.93,197,18,2381.65,134.246751
3,Website,31.595547,8509.982208,111165.14,67872.09,0.00,795,56,9383.68,139.830365


In [218]:
df_orders_channel['return_rate'] = df_orders_channel['returned_binary'] / df_orders_channel['total_orders'] * 100

In [219]:
df_orders_channel.sort_values('return_rate', ascending = False)

,channel,average_profit,profit_margin,gross_revenue,total_costs,platform_fee,total_orders,returned_binary,refund_amount,average_order_value,return_rate
2,Social Commerce,17.109340,460.431983,26446.61,18558.83,1944.93,197,18,2381.65,134.246751,9.137056
1,Mobile App,36.321392,6953.379826,82725.94,50500.09,0.00,589,43,5324.17,140.451511,7.300509
3,Website,31.595547,8509.982208,111165.14,67872.09,0.00,795,56,9383.68,139.830365,7.044025
0,Marketplace,15.397446,-1042.093842,57631.44,43053.53,7949.69,419,27,3492.95,137.545203,6.443914


In [221]:
df_orders_channel[['channel','return_rate']].sort_values('return_rate',ascending=False)

,channel,return_rate
2,Social Commerce,9.137056
1,Mobile App,7.300509
3,Website,7.044025
0,Marketplace,6.443914


In [222]:
df_orders_channel['refund_amount'].sum()

20582.449999999997

Answer: The return rate varies across sales channels, with Social Commerce showing the highest return rate at approximately 9.14%, followed by Mobile App at 7.30%, Website at 7.04%, and Marketplace at 6.44%. This suggests that customers purchasing through Social Commerce are more likely to return products, potentially due to differences in customer expectations or marketing presentation on social platforms. In contrast, Marketplace orders have the lowest return rate, indicating relatively better customer satisfaction or stricter return policies. Over the entire analysis period, the total revenue lost due to returns was estimated by summing the refund amounts associated with returned orders. The company refunded approximately $20,582.45, representing the total revenue impact of product returns during the period.

4. Analyze the marketing spend data: Which advertising platform delivers the best ROAS (Return on Ad Spend)? Are there any platforms where the company is spending money but not getting a positive return?

In [206]:
df_marketing_spend_platform.sort_values('ROAS', ascending = False)

,platform,spend,impressions,clicks,conversions,revenue_attributed,ROAS,CPC,CPA,CTR,CVR,RPC,ctr,cvr,rpc
5,TikTok Ads,57229.22,20750539,383269,18684,1374627.17,24.02,0.15,3.06,0.003,0.0487,24.02,0.018470,0.048749,3.586586
3,Influencer,97663.12,9990653,587501,28824,2216974.26,22.70,0.17,3.39,0.010,0.0491,22.70,0.058805,0.049062,3.773567
4,Instagram Ads,65154.02,11373438,306843,13863,1024639.06,15.73,0.21,4.70,0.006,0.0452,15.73,0.026979,0.045179,3.339294
2,Google Ads,152546.48,17529367,611740,32309,2194120.83,14.38,0.25,4.72,0.009,0.0528,14.38,0.034898,0.052815,3.586689
1,Facebook Ads,106451.93,16305609,331746,16968,1218572.43,11.45,0.32,6.27,0.007,0.0511,11.45,0.020346,0.051148,3.673209
0,Email Marketing,24461.37,636743,25971,1370,117681.45,4.81,0.94,17.86,0.038,0.0528,4.81,0.040787,0.052751,4.531264


Answer: Based on the marketing spend data, TikTok Ads delivers the highest ROAS, indicating that it generates the most revenue relative to the advertising budget. This suggests that TikTok Ads is the most efficient marketing channel in terms of converting advertising spend into revenue. Influencer marketing also performs strongly, with a ROAS very close to TikTok Ads, making it another highly effective platform. Instagram Ads and Google Ads show solid performance as well, generating strong returns on advertising spend and contributing significantly to revenue.
On the other hand, Email Marketing has the lowest ROAS among all platforms, meaning it generates significantly less revenue relative to the amount spent compared to other channels. However, its ROAS is still above 1, indicating that it still produces a positive return. Therefore, there are no platforms where the company is spending money without receiving a positive return, as all platforms generate more revenue than their advertising costs.

5. If the CEO asked you to cut 20% of the marketing budget, which platforms and months would you recommend reducing spend on? Support your recommendation with data.

In [207]:
df_marketing_spend_platform.sort_values('ROAS', ascending = False)

,platform,spend,impressions,clicks,conversions,revenue_attributed,ROAS,CPC,CPA,CTR,CVR,RPC,ctr,cvr,rpc
5,TikTok Ads,57229.22,20750539,383269,18684,1374627.17,24.02,0.15,3.06,0.003,0.0487,24.02,0.018470,0.048749,3.586586
3,Influencer,97663.12,9990653,587501,28824,2216974.26,22.70,0.17,3.39,0.010,0.0491,22.70,0.058805,0.049062,3.773567
4,Instagram Ads,65154.02,11373438,306843,13863,1024639.06,15.73,0.21,4.70,0.006,0.0452,15.73,0.026979,0.045179,3.339294
2,Google Ads,152546.48,17529367,611740,32309,2194120.83,14.38,0.25,4.72,0.009,0.0528,14.38,0.034898,0.052815,3.586689
1,Facebook Ads,106451.93,16305609,331746,16968,1218572.43,11.45,0.32,6.27,0.007,0.0511,11.45,0.020346,0.051148,3.673209
0,Email Marketing,24461.37,636743,25971,1370,117681.45,4.81,0.94,17.86,0.038,0.0528,4.81,0.040787,0.052751,4.531264


In [208]:
#All Spend
print(f"All Spend = {df_marketing_spend_platform['spend'].sum()}")

spend_cut = df_marketing_spend_platform['spend'].sum() * 0.2
print(f"20% Spend Cut = {spend_cut}")

print(f"20% of 20% spend cut = {(spend_cut * 0.2).round(2)}")
print(f"35% of 20% spend cut = {(spend_cut * 0.35).round(2)}")
print(f"45% of 20% spend cut = {(spend_cut * 0.45).round(2)}")

All Spend = 503506.14
20% Spend Cut = 100701.228
20% of 20% spend cut = 20140.25
35% of 20% spend cut = 35245.43
45% of 20% spend cut = 45315.55


In [209]:
# Platform Spen after 20% cut
new_cut_platform = df_marketing_spend_platform.copy()

new_cut_platform.loc[new_cut_platform['platform'] == 'Email Marketing', 'spend'] = new_cut_platform.loc[new_cut_platform['platform'] == 'Email Marketing', 'spend'] - (spend_cut * 0.2) 
new_cut_platform.loc[new_cut_platform['platform'] == 'Facebook Ads', 'spend'] = new_cut_platform.loc[new_cut_platform['platform'] == 'Facebook Ads', 'spend'] - (spend_cut * 0.35) 
new_cut_platform.loc[new_cut_platform['platform'] == 'Google Ads', 'spend'] = new_cut_platform.loc[new_cut_platform['platform'] == 'Google Ads', 'spend'] - (spend_cut * 0.45) 

new_cut_platform.sort_values('ROAS', ascending = False)

,platform,spend,impressions,clicks,conversions,revenue_attributed,ROAS,CPC,CPA,CTR,CVR,RPC,ctr,cvr,rpc
5,TikTok Ads,57229.2200,20750539,383269,18684,1374627.17,24.02,0.15,3.06,0.003,0.0487,24.02,0.018470,0.048749,3.586586
3,Influencer,97663.1200,9990653,587501,28824,2216974.26,22.70,0.17,3.39,0.010,0.0491,22.70,0.058805,0.049062,3.773567
4,Instagram Ads,65154.0200,11373438,306843,13863,1024639.06,15.73,0.21,4.70,0.006,0.0452,15.73,0.026979,0.045179,3.339294
2,Google Ads,107230.9274,17529367,611740,32309,2194120.83,14.38,0.25,4.72,0.009,0.0528,14.38,0.034898,0.052815,3.586689
1,Facebook Ads,71206.5002,16305609,331746,16968,1218572.43,11.45,0.32,6.27,0.007,0.0511,11.45,0.020346,0.051148,3.673209
0,Email Marketing,4321.1244,636743,25971,1370,117681.45,4.81,0.94,17.86,0.038,0.0528,4.81,0.040787,0.052751,4.531264


answer: To meet the CEO’s request to reduce the marketing budget by 20%, the total marketing spend was first calculated and the required reduction amount was determined as 20% of the total spend. This reduction was then distributed across selected platforms based on their performance and budget size. Email Marketing receives 20% of the total reduction because, despite its lower ROAS, it demonstrates strong engagement metrics such as high RPC, CTR, and conversion rates, suggesting potential for optimization rather than elimination. The majority of the reduction is allocated to Facebook Ads (35%) and Google Ads (45%), which represent larger portions of the marketing budget and deliver moderate performance relative to top-performing platforms such as TikTok Ads and Influencer marketing. This strategy allows the company to achieve the required budget reduction while preserving the highest-performing channels and maintaining diversified marketing coverage.

In [210]:
new_spend = []
total_roas = new_cut_platform['ROAS'].sum()
total_spend = new_cut_platform['spend'].sum()
for index, roas in enumerate(new_cut_platform['ROAS']):
    x = roas / total_roas * total_spend
    new_spend.append(x.round(2))

new_spend

[20813.1, 49544.7, 62222.95, 98223.99, 68064.47, 103935.7]

In [211]:
new_cut_platform['new_spend'] = new_spend

In [212]:
new_cut_platform.sort_values('ROAS', ascending = False)

,platform,spend,impressions,clicks,conversions,revenue_attributed,ROAS,CPC,CPA,CTR,CVR,RPC,ctr,cvr,rpc,new_spend
5,TikTok Ads,57229.2200,20750539,383269,18684,1374627.17,24.02,0.15,3.06,0.003,0.0487,24.02,0.018470,0.048749,3.586586,103935.70
3,Influencer,97663.1200,9990653,587501,28824,2216974.26,22.70,0.17,3.39,0.010,0.0491,22.70,0.058805,0.049062,3.773567,98223.99
4,Instagram Ads,65154.0200,11373438,306843,13863,1024639.06,15.73,0.21,4.70,0.006,0.0452,15.73,0.026979,0.045179,3.339294,68064.47
2,Google Ads,107230.9274,17529367,611740,32309,2194120.83,14.38,0.25,4.72,0.009,0.0528,14.38,0.034898,0.052815,3.586689,62222.95
1,Facebook Ads,71206.5002,16305609,331746,16968,1218572.43,11.45,0.32,6.27,0.007,0.0511,11.45,0.020346,0.051148,3.673209,49544.70
0,Email Marketing,4321.1244,636743,25971,1370,117681.45,4.81,0.94,17.86,0.038,0.0528,4.81,0.040787,0.052751,4.531264,20813.10


Recommendation 1: Reallocate marketing budget toward high-ROAS platforms

The marketing data shows that TikTok Ads and Influencer marketing deliver the highest ROAS, significantly outperforming other channels. To improve marketing efficiency, advertising spend should be reallocated toward these high-return platforms while reducing investment in lower ROAS channels such as Email Marketing and Facebook Ads.

Based on the ROAS-weighted allocation model, budget redistribution would increase spending on TikTok Ads and Influencer campaigns while reducing lower-efficiency platforms. This strategy maintains the same overall marketing budget while improving the expected revenue generated per dollar spent.



Recommendation 2: Reduce underperforming platform spending

Email Marketing generates the lowest ROAS, despite strong engagement metrics such as CTR and CVR. Rather than eliminating the channel entirely, the budget should be reduced while optimizing campaign design and segmentation.

For example, reducing Email Marketing spend by approximately $X would have minimal revenue impact while improving overall marketing efficiency. The savings can then be reallocated toward higher-performing channels such as TikTok Ads and Influencer campaigns.



Recommendation 3: Reduce return-related revenue losses

Product returns represent a direct loss of revenue. The analysis shows that the company lost approximately $20,582 in revenue due to refunds during the analysis period. Social Commerce has the highest return rate (approximately 9.14%), suggesting that improvements in product descriptions, customer targeting, or order fulfillment could reduce returns.